### PR TITLE
Black Market Item Stat Searching, Meso Market, and Parsing Constants

### PR DESCRIPTION
**Pull request by Zintixx**
- #68 
- Implements Black Market item stat searching
- Implements Meso Market
- Implements parsing constants.xml and also adding a json for any hardcoded/server-side only values.
- There's currently no proper way to obtain meso tokens other than the currency command.
- Added meret handling to mail. There's also an option for red merets but we're not handling that so it's being disregarded.

### CODE REVIEW

File: GameDataParser/Parsers/ConstantsParser.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+﻿using System.Xml;
+using GameDataParser.Files;
+using Maple2.File.IO.Crypto.Common;
+using Maple2Storage.Types.Metadata;
+
+namespace GameDataParser.Parsers
```
Comment (AngeloTadeucci) [inline GameDataParser/Parsers/ConstantsParser.cs]: Use file-scope namespace

File: GameDataParser/Parsers/ConstantsParser.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        protected override List<ConstantsMetadata> Parse()
+        {
+            List<ConstantsMetadata> constants = new List<ConstantsMetadata>();
+            foreach (PackFileEntry entry in Resources.XmlReader.Files)
+            {
+                if (!entry.Name.StartsWith("table/constants"))
+                {
+                    continue;
+                }
```
Comment (AngeloTadeucci) [inline GameDataParser/Parsers/ConstantsParser.cs]: Instead of looping trough every file, you could just get the first file with the desired name

File: GameDataParser/Parsers/ConstantsParser.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                {
+                    string locale = string.IsNullOrEmpty(node.Attributes["locale"]?.Value) ? "" : node.Attributes["locale"].Value;
+                    if (locale != "NA" && locale != "")
+                    {
+                        continue;
+                    }
+
+                    ConstantsMetadata metadata = new ConstantsMetadata();
+
+                    if (node.Name == "v")
```
Comment (AngeloTadeucci) [inline GameDataParser/Parsers/ConstantsParser.cs]: Invert and just use constants.Add(new(key, value))

File: GameDataParser/Parsers/ConstantsParser.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                if (!entry.Name.StartsWith("table/constants"))
+                {
+                    continue;
+                }
+
+                XmlDocument document = Resources.XmlReader.GetXmlDocument(entry);
+                foreach (XmlNode node in document.DocumentElement.ChildNodes)
+                {
+                    string locale = string.IsNullOrEmpty(node.Attributes["locale"]?.Value) ? "" : node.Attributes["locale"].Value;
+                    if (locale != "NA" && locale != "")
```
Comment (AlanMorel) [inline GameDataParser/Parsers/ConstantsParser.cs:26]: Not sure if I'm over-thinking this but if you do `if (locale != "NA"` then `locale` must be `NA` to pass this check, and if so, then you don't need to also check if `locale != ""`, right?

File: MapleServer2/Data/Static/ConstantsMetadataStorage.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
+﻿using Maple2Storage.Types;
+using Maple2Storage.Types.Metadata;
+using Newtonsoft.Json;
+using ProtoBuf;
+
+namespace MapleServer2.Data.Static
```
Comment (AngeloTadeucci) [inline MapleServer2/Data/Static/ConstantsMetadataStorage.cs]: file-scope namespace

File: MapleServer2/Packets/CharacterListPacket.cs [hunk 6]
```diff
@@ -1,6 +1,8 @@
-﻿using Maple2Storage.Enums;
+﻿
```
Comment (AlanMorel) [inline MapleServer2/Packets/CharacterListPacket.cs:1]: Unneeded extra blank line?

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: GameDataParser/Parsers/ConstantsParser.cs (hunk 4 above)
Comment (Zintixx) [reply GameDataParser/Parsers/ConstantsParser.cs:26]: It needs both. There's several shared (between regions) values with empty locale.

File: MapleServer2/PacketHandlers/Game/BlackMarketHandler.cs [hunk 7]
```diff
@@ -240,4 +262,34 @@ private static void HandlePurchase(GameSession session, PacketReader packet)
...
         session.Send(BlackMarketPacket.Purchase(listingId, amount));
     }
+
+    private static ItemStat ReadStat(int statId, int value)
+    {
+        // Normal Stat with percent value
+        if (statId >= 1000 && statId < 11000)
+        {
+            NormalStat normalStat = new NormalStat();
+            float percent = (float) Math.Round(value * 0.0001f + 0.0005f, 4);
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/BlackMarketHandler.cs:272]: why is this needed again? Can't you just cast to float the value?

File: MapleServer2/PacketHandlers/Game/Helpers/MailHelper.cs [hunk 8]
```diff
@@ -91,7 +90,52 @@ private static void SendBlackMarketSoldMail(BlackMarketListing listing, Item ite
 
         Mail mail = new(MailType.BlackMarketSale, listing.OwnerCharacterId, 0, senderName, title, body, addParameter1, addParameter2, new()
         {
-        }, revenue);
+        }, revenue, 0);
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/Helpers/MailHelper.cs:93]: the { } is not needed here right?

File: MapleServer2/PacketHandlers/Game/MesoMarketHandler.cs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+using MapleServer2.Data.Static;
+using MapleServer2.Database;
+using MapleServer2.PacketHandlers.Game.Helpers;
+using MapleServer2.Packets;
+using MapleServer2.Servers.Game;
+using MapleServer2.Types;
+
+namespace MapleServer2.PacketHandlers.Game;
+
+internal class MesoMarketHandler : GamePacketHandler
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MesoMarketHandler.cs:12]:
```suggestion
public class MesoMarketHandler : GamePacketHandler
```

File: MapleServer2/PacketHandlers/Game/MesoMarketHandler.cs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+        if (!session.Player.Wallet.Meso.Modify(-mesos))
+        {
+            session.Send(MesoMarketPacket.Error((int) MesoMarketError.NotEnoughMesosToList));
+            return;
+        }
+
+        MesoMarketListing listing = new MesoMarketListing(session.Player, price, mesos);
+
+        session.Player.Account.MesoMarketDailyListings += 1;
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MesoMarketHandler.cs:99]:
```suggestion
        session.Player.Account.MesoMarketDailyListings++;
```

File: MapleServer2/PacketHandlers/Game/MesoMarketHandler.cs [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        session.Send(MesoMarketPacket.CreateListing(listing));
+        session.Send(MesoMarketPacket.AccountStats(session.Player.Account.MesoMarketDailyListings, session.Player.Account.MesoMarketMonthlyPurchases));
+    }
+
+    private static void HandleCancelListing(GameSession session, PacketReader packet)
+    {
+        long listingId = packet.ReadLong();
+
+        MesoMarketListing listing = GameServer.MesoMarketManager.GetListingById(listingId);
+        if (listing == null)
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MesoMarketHandler.cs:110]: I would recommend you using `is` since it's the new c# standard 
```suggestion
        if (listing is null)
```

File: MapleServer2/PacketHandlers/Game/MesoMarketHandler.cs [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+            return;
+        }
+
+        if (listing.OwnerAccountId == session.Player.AccountId)
+        {
+            session.Send(MesoMarketPacket.Error((int) MesoMarketError.CantPurchaseOwnMeso));
+            return;
+        }
+
+        session.Player.Account.MesoMarketMonthlyPurchases += 1;
```
Comment (AngeloTadeucci) [inline MapleServer2/PacketHandlers/Game/MesoMarketHandler.cs:160]:
```suggestion
        session.Player.Account.MesoMarketMonthlyPurchases++;
```

File: MapleServer2/Packets/MesoMarketPacket.cs [hunk 13]
```diff
@@ -0,0 +1,119 @@
+﻿using MaplePacketLib2.Tools;
+using MapleServer2.Constants;
+using MapleServer2.Data.Static;
+using MapleServer2.Types;
+
+namespace MapleServer2.Packets
```
Comment (AngeloTadeucci) [inline MapleServer2/Packets/MesoMarketPacket.cs]: file-scope namespace

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/PacketHandlers/Game/BlackMarketHandler.cs (hunk 7 above)
Comment (Zintixx) [reply MapleServer2/PacketHandlers/Game/BlackMarketHandler.cs:272]: Without rounding it, I get values like
`Float Value: 0.0009999999`

**Review by AlanMorel — APPROVED** [review]

**Review by darkvergus — APPROVED** [review]

**Review by AngeloTadeucci — APPROVED** [review]

**Review by darkvergus — APPROVED** [review]
